### ISSUE TITLE
Images/CSS paths are wrong when using Compass and the server task

### ISSUE THREAD

**Issue by alejandroiglesias** [yeoman/generator-webapp_DEPRECATED#35]
Hi, i'm trying to build my project and this are the errors i'm finding:

Running just `grunt` works fine, gets me a working app on `dist/` dir. The problem is with the server task. When running `grunt server`, webserver seems to run from the `app/` dir and my HTML points to a stylesheet in `styles/main.css` which doesn't exists in `app/styles/` because it's a Sass file named `main.scss`. Running `grunt server:dist` seems to compile and run the server from `dist/` loading my CSS file but with wrong image paths, as it tries to fetch: `http://localhost:9000/app/images/snickers/desktop/blue-free-750.jpg` (notice `app/`).

So seems that Compass is getting it wrong. I'm using the great Compass image-url() helper to load images like:

``` css
background-image: image-url('snickers/desktop/blue-free-750.jpg');
```

Gruntfile Compass options are the defaults:

``` javascript
options: {
    sassDir: '<%= yeoman.app %>/styles',
    cssDir: '.tmp/styles',
    imagesDir: '<%= yeoman.app %>/images',
    javascriptsDir: '<%= yeoman.app %>/scripts',
    fontsDir: '<%= yeoman.app %>/styles/fonts',
    importPath: 'app/components',
    relativeAssets: true
},
```

If i change `imagesDir` value to just `'images'`, it works when running `grunt server:dist` but then running `grunt` and opening from `dist/` doesn't, fails to load the image again (`file://localhost/Users/alejandro/Development/nike-free/images/snickers/desktop/blue-free-750.jpg`).
Removing the image-url() helper from the source couldn't make it work running `grunt`, `grunt server` nor `grunt server:dist`.

Next is the output of my environment by running the command `yo --version && echo $PATH $NODE_PATH && node -e 'console.log(process.platform, process.versions)' && cat Gruntfile.js`:

``` javascript
1.0.0-beta.3
/Users/alejandro/.rvm/gems/ruby-1.9.3-p194/bin:/Users/alejandro/.rvm/gems/ruby-1.9.3-p194@global/bin:/Users/alejandro/.rvm/rubies/ruby-1.9.3-p194/bin:/Users/alejandro/.rvm/bin:/usr/local/bin:/usr/local/share/npm/bin:/usr/local/share/npm/lib:/Users/alejandro/.rvm/bin:/var/lib/gems/1.8/bin:/Users/alejandro/.gem/ruby/1.8/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/X11/bin
darwin { http_parser: '1.0',
  node: '0.8.20',
  v8: '3.11.10.25',
  ares: '1.7.5-DEV',
  uv: '0.8',
  zlib: '1.2.3',
  openssl: '1.0.0f' }
// Generated on 2013-03-19 using generator-webapp 0.1.5
'use strict';
var lrSnippet = require('grunt-contrib-livereload/lib/utils').livereloadSnippet;
var mountFolder = function (connect, dir) {
    return connect.static(require('path').resolve(dir));
};

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// use this if you want to match all subfolders:
// 'test/spec/**/*.js'

module.exports = function (grunt) {
    // load all grunt tasks
    require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

    // configurable paths
    var yeomanConfig = {
        app: 'app',
        dist: 'dist'
    };

    grunt.initConfig({
        yeoman: yeomanConfig,
        watch: {
            compass: {
                files: ['<%= yeoman.app %>/styles/{,*/}*.{scss,sass}'],
                tasks: ['compass']
            },
            livereload: {
                files: [
                    '<%= yeoman.app %>/*.html',
                    '{.tmp,<%= yeoman.app %>}/styles/{,*/}*.css',
                    '{.tmp,<%= yeoman.app %>}/scripts/{,*/}*.js',
                    '<%= yeoman.app %>/images/{,*/}*.{png,jpg,jpeg,webp}'
                ],
                tasks: ['livereload']
            }
        },
        connect: {
            options: {
                port: 9000,
                // change this to '0.0.0.0' to access the server from outside
                hostname: 'localhost'
            },
            livereload: {
                options: {
                    middleware: function (connect) {
                        return [
                            lrSnippet,
                            mountFolder(connect, '.tmp'),
                            mountFolder(connect, 'app')
                        ];
                    }
                }
            },
            test: {
                options: {
                    middleware: function (connect) {
                        return [
                            mountFolder(connect, '.tmp'),
                            mountFolder(connect, 'test')
                        ];
                    }
                }
            },
            dist: {
                options: {
                    middleware: function (connect) {
                        return [
                            mountFolder(connect, 'dist')
                        ];
                    }
                }
            }
        },
        open: {
            server: {
                path: 'http://localhost:<%= connect.options.port %>'
            }
        },
        clean: {
            dist: ['.tmp', '<%= yeoman.dist %>/*'],
            server: '.tmp'
        },
        jshint: {
            options: {
                jshintrc: '.jshintrc'
            },
            all: [
                'Gruntfile.js',
                '<%= yeoman.app %>/scripts/{,*/}*.js',
                '!<%= yeoman.app %>/scripts/vendor/*',
                'test/spec/{,*/}*.js'
            ]
        },
        mocha: {
            all: {
                options: {
                    run: true,
                    urls: ['http://localhost:<%= connect.options.port %>/index.html']
                }
            }
        },
        compass: {
            options: {
                sassDir: '<%= yeoman.app %>/styles',
                cssDir: '.tmp/styles',
                imagesDir: '<%= yeoman.app %>/images',
                javascriptsDir: '<%= yeoman.app %>/scripts',
                fontsDir: '<%= yeoman.app %>/styles/fonts',
                importPath: 'app/components',
                relativeAssets: true
            },
            dist: {},
            server: {
                options: {
                    debugInfo: true
                }
            }
        },
        // not used since Uglify task does concat,
        // but still available if needed
        /*concat: {
            dist: {}
        },*/
        requirejs: {
            dist: {
                // Options: https://github.com/jrburke/r.js/blob/master/build/example.build.js
                options: {
                    // `name` and `out` is set by grunt-usemin
                    baseUrl: 'app/scripts',
                    optimize: 'none',
                    // TODO: Figure out how to make sourcemaps work with grunt-usemin
                    // https://github.com/yeoman/grunt-usemin/issues/30
                    //generateSourceMaps: true,
                    // required to support SourceMaps
                    // http://requirejs.org/docs/errors.html#sourcemapcomments
                    preserveLicenseComments: false,
                    useStrict: true,
                    wrap: true,
                    //uglify2: {} // https://github.com/mishoo/UglifyJS2
                }
            }
        },
        useminPrepare: {
            html: '<%= yeoman.app %>/index.html',
            options: {
                dest: '<%= yeoman.dist %>'
            }
        },
        usemin: {
            html: ['<%= yeoman.dist %>/{,*/}*.html'],
            css: ['<%= yeoman.dist %>/styles/{,*/}*.css'],
            options: {
                dirs: ['<%= yeoman.dist %>']
            }
        },
        imagemin: {
            dist: {
                files: [{
                    expand: true,
                    cwd: '<%= yeoman.app %>/images',
                    src: '**/*.{png,jpg,jpeg}',
                    dest: '<%= yeoman.dist %>/images'
                }]
            }
        },
        cssmin: {
            dist: {
                files: {
                    '<%= yeoman.dist %>/styles/main.css': [
                        '.tmp/styles/{,*/}*.css',
                        '<%= yeoman.app %>/styles/{,*/}*.css'
                    ]
                }
            }
        },
        htmlmin: {
            dist: {
                options: {
                    /*removeCommentsFromCDATA: true,
                    // https://github.com/yeoman/grunt-usemin/issues/44
                    //collapseWhitespace: true,
                    collapseBooleanAttributes: true,
                    removeAttributeQuotes: true,
                    removeRedundantAttributes: true,
                    useShortDoctype: true,
                    removeEmptyAttributes: true,
                    removeOptionalTags: true*/
                },
                files: [{
                    expand: true,
                    cwd: '<%= yeoman.app %>',
                    src: '*.html',
                    dest: '<%= yeoman.dist %>'
                }]
            }
        },
        copy: {
            dist: {
                files: [{
                    expand: true,
                    dot: true,
                    cwd: '<%= yeoman.app %>',
                    dest: '<%= yeoman.dist %>',
                    src: [
                        '*.{ico,txt}',
                        '.htaccess'
                    ]
                }]
            }
        },
        bower: {
            all: {
                rjsConfig: '<%= yeoman.app %>/scripts/main.js'
            }
        }
    });

    grunt.renameTask('regarde', 'watch');

    grunt.registerTask('server', function (target) {
        if (target === 'dist') {
            return grunt.task.run(['build', 'open', 'connect:dist:keepalive']);
        }

        grunt.task.run([
            'clean:server',
            'compass:server',
            'livereload-start',
            'connect:livereload',
            'open',
            'watch'
        ]);
    });

    grunt.registerTask('test', [
        'clean:server',
        'compass',
        'connect:test',
        'mocha'
    ]);

    grunt.registerTask('build', [
        'clean:dist',
        'compass:dist',
        'useminPrepare',
        'requirejs',
        'imagemin',
        'htmlmin',
        'concat',
        'cssmin',
        'uglify',
        'copy',
        'usemin'
    ]);

    grunt.registerTask('default', [
        'jshint',
        'test',
        'build'
    ]);
};
```

**Comment by kevva** [comment]
You could use the compass option `http_images_path`. There's no option for that in the compass task but if you read the docs there are some workarounds.

> Compass doesn't expose all of its options through the CLI, which this task makes use of. If you need an option not mentioned below you can either specify a path to a config.rb file in the config option or embed it directly into the raw option. Options defined in your Gruntfile will override those specified in your config.rb or raw property. config and raw are mutually exclusive.

Something like this could work:

``` js
compass: {
  options: {
    raw: 'http_images_path = \"/images\"'
  }
}
```

Maybe this is something we should include by default since any images generated through compass will always point to app/images? cc @sindresorhus @passy

**Comment by alejandroiglesias** [comment]
@kevva thanks for your reply. What you suggested worked partially. I also had to set `relativeAssets` to `false` for this to work with `grunt server:dist`. But still the problem is i need `grunt server` to work for development.
One progress on this is that i see that when running `grunt server` the webserver mounts two folders: `.tmp/` and `app/`. My stylesheets in `app/styles/` are in Sass format, and they're as CSS on `.tmp/` but not under the name of `main.css` as i'm referencing in my HTML file but as separated files (one CSS file for each SCSS file, even an empty `mixins.css` file).
So to summarize:
- When running `grunt server`, webserver mounts `.tmp/` and `app/` and my stylesheets are in `.tmp/` but with the name of the original separated files.
- When running `grunt` or `grunt server:dist` my stylesheets are in `dist/` concatenated into `main.css`, which is what i'm referencing in my HTML files.

So how do i get a `.tmp/main.css` file?

**Comment by alejandroiglesias** [comment]
I also want to add that using:

``` javascript
options: {
    relativeAssets: false,
    raw: 'http_images_path = \"/images\"'
}
```

Made running `grunt` alone and opening the file without webserver fail to load images.

**Comment by felquis** [comment]
@kevva It's a bug of the webapp?

**Comment by kevva** [comment]
Please can you describe the problem more exactly? What isn't working? Compass images? CSS in `.tmp`?

**Comment by felquis** [comment]
The problem is this:

In main.scss I use

``` css
body {
    background: image-url('my-image.jpg');
}
```

And the output is

``` css
body {
    background: url('/app/images/my-image.jpg?12134343');
}
```

But, the correct output would be:

``` css
body {
    background: url('/images/my-image.jpg?12134343');
}
```

This solve the problem

```
raw: 'http_images_path = \"/images\"'
```

But this is the correct way to fix this?

**Comment by kevva** [comment]
Yes. Or if you don't use the Compass helper `image-url()` :

``` css
background: url("../images/my-image.jpg");
```

**Comment by passy** [comment]
You don't have to use `raw` for those options anymore since 0.2.0: https://github.com/gruntjs/grunt-contrib-compass#README

**Comment by felquis** [comment]
Using sprite generator 

![yo-compass](https://f.cloud.github.com/assets/736728/412479/a591fd7e-abb7-11e2-9c58-d20ebc5c9f4d.PNG)

@passy I updated the package to v0.2.0 but doesn't fixed the problem.

@kevva You know how I can fix it?

**Comment by felquis** [comment]
I fixed it using

``` js
raw: 'http_images_path = \"/images\"; http_generated_images_path = \"/images\"'
```

**Comment by kevva** [comment]
Didn't `httpImagesPath: '/images', httpGeneratedImagesPath: '/images'` work with 0.2.0?

**Comment by passy** [comment]
This is no longer an issue with `grunt-contrib-compass` >= 0.3.0
